core/src/main/java/com/scalar/db/common/AbstractDistributedTransactionManager.java

```diff
@@ -76,151 +71,159 @@ public Optional<String> getTable() {
-    public void put(List<Put> puts) throws CrudException {
-      checkIfActive();
-      super.put(puts);
-    }
+  @Override
+  public void delete(Delete delete) throws CrudException, UnknownTransactionStatusException {
+    executeTransaction(
+        t -> {
+          t.delete(copyAndSetTargetToIfNot(delete));
+          return null;
+        });
+  }
 
-    @Override
-    public void delete(Delete delete) throws CrudException {
-      checkIfActive();
-      super.delete(delete);
-    }
+  @Deprecated
+  @Override
+  public void delete(List<Delete> deletes) throws CrudException, UnknownTransactionStatusException {
+    executeTransaction(
+        t -> {
+          t.delete(copyAndSetTargetToIfNot(deletes));
+          return null;
+        });
+  }
 
-    /** @deprecated As of release 3.13.0. Will be removed in release 5.0.0. */
-    @Deprecated
-    @Override
-    public void delete(List<Delete> deletes) throws CrudException {
-      checkIfActive();
-      super.delete(deletes);
-    }
+  @Override
+  public void mutate(List<? extends Mutation> mutations)
+      throws CrudException, UnknownTransactionStatusException {
+    executeTransaction(
+        t -> {
+          t.mutate(copyAndSetTargetToIfNot(mutations));
+          return null;
+        });
+  }
 
-    @Override
-    public void insert(Insert insert) throws CrudException {
-      checkIfActive();
-      super.insert(insert);
+  private <R> R executeTransaction(
+      ThrowableFunction<DistributedTransaction, R, TransactionException> throwableFunction)
+      throws CrudException, UnknownTransactionStatusException {
+    DistributedTransaction transaction;
+    try {
+      transaction = begin();
+    } catch (TransactionNotFoundException e) {
+      throw new CrudConflictException(e.getMessage(), e, e.getTransactionId().orElse(null));
+    } catch (TransactionException e) {
+      throw new CrudException(e.getMessage(), e, e.getTransactionId().orElse(null));
+    }
+
+    try {
+      R result = throwableFunction.apply(transaction);
+      transaction.commit();
+      return result;
+    } catch (CrudException e) {
+      rollbackTransaction(transaction);
+      throw e;
+    } catch (CommitConflictException e) {
+      rollbackTransaction(transaction);
+      throw new CrudConflictException(e.getMessage(), e, e.getTransactionId().orElse(null));
+    } catch (UnknownTransactionStatusException e) {
+      throw e;
+    } catch (TransactionException e) {
+      rollbackTransaction(transaction);
+      throw new CrudException(e.getMessage(), e, e.getTransactionId().orElse(null));
     }
+  }
 
-    @Override
-    public void upsert(Upsert upsert) throws CrudException {
-      checkIfActive();
-      super.upsert(upsert);
+  private void rollbackTransaction(DistributedTransaction transaction) throws CrudException {
+    try {
+      transaction.rollback();
+    } catch (RollbackException e) {
+      throw new CrudException(e.getMessage(), e, e.getTransactionId().orElse(null));
```

Comment:
Uncommitted transaction will be rolled back, so just warning message or something is enough here?

Reply:
Correct. Let me add a warning message here. Thanks.

Reply:
Fixed in 0b1c67ce83a356f118a5e8ab635a716fd563413b. Thanks.